File: smoke_tests/swift/ldap_injection.swift
// LDAP Injection vulnerabilities in Swift
import Foundation

class LdapInjectionVulnerabilities {

    let ldapHost = "ldap://directory.example.com"

    // Test 1: User search with unescaped input
    func findUser(username: String) -> String {
        // VULNERABLE: Username directly in filter
        let filter = "(uid=\(username))"
        return executeLdapQuery(filter: filter)
    }

    // Test 2: Authentication with LDAP
    func authenticateUser(username: String, password: String) -> Bool {
        // VULNERABLE: Credentials in LDAP bind
        let dn = "uid=\(username),ou=users,dc=example,dc=com"
        return ldapBind(dn: dn, password: password)
    }

    // Test 3: Group membership check
    func isUserInGroup(username: String, group: String) -> Bool {
        // VULNERABLE: Both parameters unescaped
        let filter = "(&(uid=\(username))(memberOf=cn=\(group),ou=groups,dc=example,dc=com))"
        return !executeLdapQuery(filter: filter).isEmpty
    }

    // Test 4: Email lookup
    func findByEmail(email: String) -> [String] {
        // VULNERABLE: Email in filter
        let filter = "(mail=\(email))"
        return executeLdapSearch(filter: filter)
    }

    // Test 5: Complex search with multiple fields
    func searchUsers(firstName: String, lastName: String, department: String) -> [String] {
        // VULNERABLE: Multiple unescaped fields
        let filter = "(&(givenName=\(firstName))(sn=\(lastName))(department=\(department)))"
        return executeLdapSearch(filter: filter)
    }

    // Test 6: Wildcard search
    func searchByPartialName(partial: String) -> [String] {
        // VULNERABLE: Wildcard with user input
        let filter = "(cn=*\(partial)*)"
        return executeLdapSearch(filter: filter)
    }

    // Test 7: DN construction
    func getUserDn(username: String, orgUnit: String) -> String {
        // VULNERABLE: DN injection
        return "uid=\(username),ou=\(orgUnit),dc=example,dc=com"
    }

    // Test 8: Role-based query
    func getUsersWithRole(role: String) -> [String] {
        // VULNERABLE: Role from user input
        let filter = "(role=\(role))"
        return executeLdapSearch(filter: filter)
    }

    // Test 9: Attribute value injection
    func findByAttribute(attribute: String, value: String) -> [String] {
        // VULNERABLE: Both attribute and value from user
        let filter = "(\(attribute)=\(value))"
        return executeLdapSearch(filter: filter)
    }

    // Test 10: Nested group query
    func findNestedGroupMembers(groupName: String) -> [String] {
        // VULNERABLE: Group name injection
        let filter = "(memberOf:1.2.840.113556.1.4.1941:=cn=\(groupName),ou=groups,dc=example,dc=com)"
        return executeLdapSearch(filter: filter)
    }

    // Test 11: User modification
    func updateUserAttribute(username: String, attribute: String, value: String) {
        // VULNERABLE: All parameters from user
        let dn = "uid=\(username),ou=users,dc=example,dc=com"
        ldapModify(dn: dn, attribute: attribute, value: value)
    }

    // Test 12: Search base manipulation
    func searchInOu(ouName: String, filter: String) -> [String] {
        // VULNERABLE: OU from user
        let searchBase = "ou=\(ouName),dc=example,dc=com"
        return executeLdapSearchWithBase(base: searchBase, filter: filter)
    }

    // Helper functions (stubs)
    private func executeLdapQuery(filter: String) -> String { "" }
    private func executeLdapSearch(filter: String) -> [String] { [] }
    private func executeLdapSearchWithBase(base: String, filter: String) -> [String] { [] }
    private func ldapBind(dn: String, password: String) -> Bool { false }
    private func ldapModify(dn: String, attribute: String, value: String) {}
}

File: smoke_tests/rust/ldap_injection.rs
// LDAP Injection Test Cases

// Test 1: LDAP search filter with unsanitized user input
fn search_user_by_username(username: &str) -> String {
    // VULNERABLE: username could contain LDAP special characters like *, (, ), \, etc.
    format!("(uid={})", username)
}

// Test 2: LDAP authentication filter
fn create_auth_filter(username: &str, password: &str) -> String {
    // VULNERABLE: username could be "admin)(&(1=1" to bypass auth
    format!("(&(uid={})(password={}))", username, password)
}

// Test 3: Complex LDAP filter with user input
fn find_users_filter(search_term: &str, department: &str) -> String {
    // VULNERABLE: Both searchTerm and department are unsanitized
    format!("(&(cn=*{}*)(department={}))", search_term, department)
}

// Test 4: LDAP DN construction
fn get_user_dn(username: &str, domain: &str) -> String {
    // VULNERABLE: username could contain DN special characters
    format!("cn={},dc={},dc=com", username, domain)
}

// Test 5: LDAP group membership filter
fn check_group_membership(username: &str, group: &str) -> String {
    // VULNERABLE: Both parameters unsanitized
    format!("(&(member=uid={},ou=users,dc=example,dc=com)(cn={}))", username, group)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_ldap_injection() {
        let filter = search_user_by_username("admin");
        assert!(filter.contains("admin"));
    }
}
